File: practice_js/problem_spec_ruby/08-two-sum.rb
# Write a method that takes an array of numbers. If a pair of numbers
# in the array sums to zero, return the positions of those two numbers.
# If no pair of numbers sums to zero, return `nil`.
#
# Difficulty: medium.

def two_sum(nums)
  hash = load_hash(nums)
  ans = []
  nums.each_with_index do |number,index|
    search_item = 0 - number
    if hash[search_item]
      return [index, nums.index(search_item)]
    end
  end
  return nil
  
end

def load_hash(nums)
  hash = nums.each_with_object({}) {|num, hash| hash[num] = true}
  return hash
end

# These are tests to check that your code is working. After writing
# your solution, they should all print true.

puts(
  'two_sum([1, 3, 5, -3]) == [1, 3]: ' + (two_sum([1, 3, 5, -3]) == [1, 3]).to_s
)
puts(
  'two_sum([1, 3, 5]) == nil: ' + (two_sum([1, 3, 5]) == nil).to_s
)

File: practice_js/javascript_solutions/jasmine-standalone-2.5.2/src/08-two-sum.js
// # Write a method that takes an array of numbers. If a pair of numbers
// # in the array sums to zero, return the positions of those two numbers.
// # If no pair of numbers sums to zero, return `nil`.

function twoSum(array){
    var hash = makeObject(array);
    var hashKeys = Object.keys(hash);
    var found = [];
    hashKeys.forEach(function(key){
        var search_item = 0 - parseInt(key)
        if(hash.hasOwnProperty(search_item) && found.length == 0){
            found.push(array.indexOf(parseInt(key)));
            found.push(array.indexOf(search_item));
        }
    });
    return found;
}

function makeObject(arr){
    var hash = {};
    arr.forEach(function(element){
        if (hash[element]){
            hash[element] += 1;
        }else{
            hash[element] = 1;
        }
    });
    return hash;
}

